File: sdks/go/docflow/markdown.go
// Package docflow provides a standalone library for converting between
// Markdown and PDF formats without requiring a server.
package docflow

import (
	"bytes"
	"sort"
	"strings"

	"github.com/yuin/goldmark"
	"github.com/yuin/goldmark/extension"
	"github.com/yuin/goldmark/parser"
	"github.com/yuin/goldmark/renderer/html"
)

// MarkdownParser handles markdown processing.
type MarkdownParser struct {
	md goldmark.Markdown
}

// NewMarkdownParser creates a new markdown parser with sensible defaults.
func NewMarkdownParser() *MarkdownParser {
	md := goldmark.New(
		goldmark.WithExtensions(
			extension.GFM,
			extension.Table,
			extension.Strikethrough,
			extension.TaskList,
			extension.Footnote,
			extension.DefinitionList,
		),
		goldmark.WithParserOptions(
			parser.WithAutoHeadingID(),
		),
		goldmark.WithRendererOptions(
			html.WithHardWraps(),
			html.WithXHTML(),
			html.WithUnsafe(),
		),
	)

	return &MarkdownParser{md: md}
}

// ToHTML converts markdown content to HTML.
func (p *MarkdownParser) ToHTML(content string) (string, error) {
	var buf bytes.Buffer
	if err := p.md.Convert([]byte(content), &buf); err != nil {
		return "", err
	}
	return buf.String(), nil
}

// MergeFiles merges multiple files into a single content string.
// Files are sorted by their Order field.
func (p *MarkdownParser) MergeFiles(files []MDFile) string {
	if len(files) == 0 {
		return ""
	}

	// Sort files by order
	sortedFiles := make([]MDFile, len(files))
	copy(sortedFiles, files)
	sort.Slice(sortedFiles, func(i, j int) bool {
		return sortedFiles[i].Order < sortedFiles[j].Order
	})

	var builder strings.Builder
	for i, file := range sortedFiles {
		if i > 0 {
			// Add page break marker and separator between files
			builder.WriteString("\n\n---\n\n")
		}
		builder.WriteString(file.Content)
	}

	return builder.String()
}

// MergeFilesToHTML merges files and converts to HTML with file separators.
func (p *MarkdownParser) MergeFilesToHTML(files []MDFile) (string, error) {
	if len(files) == 0 {
		return "", nil
	}

	// Sort files by order
	sortedFiles := make([]MDFile, len(files))
	copy(sortedFiles, files)
	sort.Slice(sortedFiles, func(i, j int) bool {
		return sortedFiles[i].Order < sortedFiles[j].Order
	})

	var builder strings.Builder
	for i, file := range sortedFiles {
		if i > 0 {
			builder.WriteString(`<div class="file-separator"><span>`)
			builder.WriteString(file.Name)
			builder.WriteString(`</span></div>`)
		} else {
			builder.WriteString(`<div class="file-header"><span>`)
			builder.WriteString(file.Name)
			builder.WriteString(`</span></div>`)
		}

		html, err := p.ToHTML(file.Content)
		if err != nil {
			return "", err
		}
		builder.WriteString(`<div class="file-content">`)
		builder.WriteString(html)
		builder.WriteString(`</div>`)
	}

	return builder.String(), nil
}

// EstimatePageCount estimates the number of PDF pages based on content.
func (p *MarkdownParser) EstimatePageCount(content string) int {
	// Rough estimation: ~3000 characters per page
	const charsPerPage = 3000
	pages := len(content) / charsPerPage
	if pages < 1 {
		return 1
	}
	return pages
}

File: sdks/python/docflow/markdown.py
"""Markdown parser for DocFlow."""

from typing import List, Optional

from .types import MDFile


class MarkdownParser:
    """Handles markdown processing and HTML conversion.
    
    Uses markdown-it-py for parsing with extensions support.
    
    Example:
        >>> parser = MarkdownParser()
        >>> html = parser.to_html("# Hello World")
    """

    def __init__(self) -> None:
        """Initialize the markdown parser."""
        try:
            from markdown_it import MarkdownIt
            from mdit_py_plugins.tasklists import tasklists_plugin
            from mdit_py_plugins.footnote import footnote_plugin
        except ImportError:
            raise ImportError(
                "markdown-it-py is required: pip install markdown-it-py mdit-py-plugins"
            )

        self.md = (
            MarkdownIt("gfm-like")
            .enable("table")
            .enable("strikethrough")
            .use(tasklists_plugin)
            .use(footnote_plugin)
        )

    def to_html(self, content: str) -> str:
        """Convert markdown content to HTML.
        
        Args:
            content: Markdown string.
            
        Returns:
            HTML string.
        """
        return self.md.render(content)

    def merge_files(self, files: List[MDFile]) -> str:
        """Merge multiple files into a single content string.
        
        Files are sorted by their order field.
        
        Args:
            files: List of MDFile objects.
            
        Returns:
            Merged markdown content.
        """
        if not files:
            return ""

        # Sort files by order
        sorted_files = sorted(files, key=lambda f: f.order)

        parts = []
        for i, file in enumerate(sorted_files):
            if i > 0:
                parts.append("\n\n---\n\n")
            parts.append(file.content)

        return "".join(parts)

    def merge_files_to_html(self, files: List[MDFile]) -> str:
        """Merge files and convert to HTML with file separators.
        
        Args:
            files: List of MDFile objects.
            
        Returns:
            HTML string with file separators.
        """
        if not files:
            return ""

        sorted_files = sorted(files, key=lambda f: f.order)

        parts = []
        for i, file in enumerate(sorted_files):
            if i > 0:
                parts.append(f'<div class="file-separator"><span>{file.name}</span></div>')
            else:
                parts.append(f'<div class="file-header"><span>{file.name}</span></div>')

            html = self.to_html(file.content)
            parts.append(f'<div class="file-content">{html}</div>')

        return "".join(parts)

    def estimate_page_count(self, content: str) -> int:
        """Estimate the number of PDF pages based on content.
        
        Args:
            content: Markdown or text content.
            
        Returns:
            Estimated page count (minimum 1).
        """
        chars_per_page = 3000
        pages = len(content) // chars_per_page
        return max(1, pages)
